crates/block-producer/src/block_producer.rs

```diff
@@ -244,83 +249,114 @@ impl BlockProducer {
+            log::info!(
+                "produce new block #{} (txs: {}, deposits: {}, withdrawals: {})",
+                number,
+                block_txs,
+                deposit_cells.len(),
+                block_withdrawals,
+            );
+            if !block.withdrawals().is_empty() && opt_finalized_custodians.is_none() {
+                bail!("unexpected none custodians for withdrawals ",);
+            }
+            let finalized_custodians = opt_finalized_custodians.unwrap_or_default();
 
-        if let Some(ref tests_control) = self.tests_control {
-            if let Some(TestModePayload::BadBlock { .. }) = tests_control.payload().await {
-                let (bad_block, bad_global_state) = tests_control
-                    .generate_a_bad_block(block, global_state)
-                    .await?;
+            if let Some(ref tests_control) = self.tests_control {
+                if let Some(TestModePayload::BadBlock { .. }) = tests_control.payload().await {
+                    let (bad_block, bad_global_state) = tests_control
+                        .generate_a_bad_block(block, global_state)
+                        .await?;
 
-                block = bad_block;
-                global_state = bad_global_state;
+                    block = bad_block;
+                    global_state = bad_global_state;
+                }
             }
-        }
 
-        // composite tx
-        let tx = match self
-            .complete_tx_skeleton(
-                deposit_cells,
-                finalized_custodians,
-                block,
-                global_state,
-                median_time,
-                rollup_cell,
-            )
-            .await
-        {
-            Ok(tx) => tx,
-            Err(err) => {
-                log::error!(
-                    "[produce_next_block] Failed to composite submitting transaction: {}",
-                    err
-                );
-                self.mem_pool.lock().await.reset_mem_block()?;
-                return Err(err);
+            // composite tx
+            let tx = match self
+                .complete_tx_skeleton(
+                    deposit_cells,
+                    finalized_custodians,
+                    block,
+                    global_state,
+                    median_time,
+                    rollup_cell.clone(),
+                )
+                .await
+            {
+                Ok(tx) => tx,
+                Err(err) => {
+                    log::error!(
+                        "[produce_next_block] Failed to composite submitting transaction: {}",
+                        err
+                    );
+                    self.mem_pool.lock().await.reset_mem_block()?;
+                    return Err(err);
+                }
+            };
+
+            if tx.as_slice().len() > MAX_BLOCK_BYTES as usize {
+                // Drop some percentage of withdrawals and txs
+                let max_withdrawals = block_withdrawals
```

Comment:
We should try to drop txs first, then only to drop withdrawals when there are no tx get packaged into the block.

And we can adjust the strategy to drop half of the txs(or withdrawals) each time.

Reply:
Fixed